win32/src/srw_lock_win32.c

```diff
@@ -212,6 +212,59 @@ void srw_lock_acquire_exclusive(SRW_LOCK_HANDLE handle)
     }
 }
 
+int srw_lock_try_acquire_exclusive(SRW_LOCK_HANDLE handle)
+{
+    int result;
+
+    if (handle == NULL)
+    {
+        /* Codes_SRS_SRW_LOCK_01_006: [ If handle is NULL then srw_lock_try_acquire_exclusive shall fail and return a non-zero value. ]*/
+        LogError("invalid argument SRW_LOCK_HANDLE handle=%p", handle);
+        result = MU_FAILURE;
+    }
+    else
+    {
+        LARGE_INTEGER start = { 0 }, stop;
+
+        if (handle->doStatistics)
+        {
+            (void)QueryPerformanceCounter(&start);
+        }
+
+        /*Codes_SRS_SRW_LOCK_01_007: [ Otherwise srw_lock_acquire_exclusive shall call TryAcquireSRWLockExclusive. ]*/
+        if (!TryAcquireSRWLockExclusive(&handle->lock))
+        {
+            /* Codes_SRS_SRW_LOCK_01_008: [ If TryAcquireSRWLockExclusive returns FALSE, srw_lock_acquire_exclusive shall return a non-zero value. ]*/
+            result = MU_FAILURE;
+        }
+        else
+        {
+            /*Codes_SRS_SRW_LOCK_01_010: [ If do_statistics is true and if the timer created has recorded more than TIME_BETWEEN_STATISTICS_LOG seconds then statistics will be logged and the timer shall be started again. ]*/
+            if (handle->doStatistics)
+            {
+                (void)QueryPerformanceCounter(&stop); /*measure acquire time*/
+
+                (void)InterlockedIncrement64(&handle->nCalls_AcquireSRWLockExclusive);
+                (void)InterlockedAdd64(&handle->totalCounts_AcquireSRWLockExclusive, stop.QuadPart - start.QuadPart);
+
+                /*piggyback on the lock itself to print the statistics "until now" - note - does not include the current Lock call*/
+                if (timer_get_elapsed(handle->timer) >= TIME_BETWEEN_STATISTICS_LOG)
+                {
+                    timer_start(handle->timer);
+                    LogStatistics(handle, "periodic logging almost every " MU_TOSTRING(TIME_BETWEEN_STATISTICS_LOG) " seconds");
+                }
+
+                (void)InterlockedExchange64(&handle->lastCount_AcquireSRWLockExclusive, stop.QuadPart);
+            }
```

Comment:
Suggest moving these (the shared and exclusive versions of statistics) to helpers instead of copy-pasting

Reply:
I like this suggestion quite a lot, applied.

---
In reply to: [461957918](https://github.com/Azure/azure-c-pal/pull/20#discussion_r461957918) [](ancestors = 461957918)